### ISSUE TITLE
[Failing Test] gce-master-scale-performance - RC timed out

### ISSUE THREAD

**Issue by oxddr** [kubernetes/kubernetes#79711]
**Which jobs are failing**:
gce-master-scale-performance

**Which test(s) are failing**:
load

**Since when has it been failing**:
02-07-2019

**Testgrid link**:
https://k8s-testgrid.appspot.com/sig-scalability-gce#gce-master-scale-performance

**Reason for failure**:
RC has timeout out. The reason for the timeout seems to be related to slow API calls. 

**Anything else we need to know**:
https://prow.k8s.io/view/gcs/kubernetes-jenkins/logs/ci-kubernetes-e2e-gce-scale-performance/1145965714423681025

/sig scalability
/assign @oxddr

**Comment by oxddr** [comment]
Actual error:

```
--------------------------------------------------------------------------------                                                                                                                                                              
 Test Finished
   Test: testing/load/config.yaml
   Status: Fail
   Errors: [measurement call WaitForControlledPodsRunning - WaitForRunningRCs error: 1 objects timed out: ReplicationControllers: test-u3fcbg-30/small-rc-39]                                                                                  
--------------------------------------------------------------------------------                                                                             
```

All mentions of test-u3fcbg-30/small-rc-39 
```
$ fgrep  small-rc-39 build-log.txt | fgrep test-u3fcbg-30                                                                                                                                                                                                               
W0702 13:11:26.917] I0702 13:11:26.917301   14684 wait_for_pods.go:141] WaitForControlledPodsRunning: namespace(test-u3fcbg-30), labelSelector(name=small-rc-39): Pods: 5 out of 5 created, 3 running, 2 pending scheduled, 0 not scheduled, 0 inactive, 0 terminating, 0 unknown, 0 runningButNotReady                        
W0702 13:11:31.917] I0702 13:11:31.917591   14684 wait_for_pods.go:141] WaitForControlledPodsRunning: namespace(test-u3fcbg-30), labelSelector(name=small-rc-39): Pods: 5 out of 5 created, 5 running, 0 pending scheduled, 0 not scheduled, 0 inactive, 0 terminating, 0 unknown, 0 runningButNotReady                        
W0702 14:34:05.394] E0702 14:34:05.394131   14684 wait_for_controlled_pods.go:495] WaitForControlledPodsRunning: error for test-u3fcbg-30/small-rc-39: pod store creation error: timed out waiting for the condition                                                                                                           
W0702 14:34:15.614] I0702 14:34:15.612925   14684 trace.go:81] Trace[1512726023]: "Reflector PodStore: namespace(test-u3fcbg-30), labelSelector(name=small-rc-39) ListAndWatch" (started: 2019-07-02 14:32:05.393780229 +0000 UTC m=+22640.906973464) (total time: 2m10.21911428s):                                            
W0702 15:32:25.644] E0702 15:32:25.644496   14684 wait_for_controlled_pods.go:285] Timed out ReplicationControllers: test-u3fcbg-30/small-rc-39                                                                                                                                                                                
W0702 16:19:05.702] W0702 16:19:05.701965   14684 reflector.go:303] PodStore: namespace(test-u3fcbg-30), labelSelector(name=small-rc-39): watch of *v1.Pod ended with: too old resource version: 17563317 (17657940)                                                                                                           
W0702 17:26:43.944] I0702 17:26:43.944552   14684 wait_for_pods.go:141] WaitForControlledPodsRunning: namespace(test-u3fcbg-30), labelSelector(name=small-rc-39): Pods: 0 out of 0 created, 0 running, 0 pending scheduled, 0 not scheduled, 0 inactive, 3 terminating, 0 unknown, 0 runningButNotReady                        
W0702 17:26:48.945] I0702 17:26:48.944951   14684 wait_for_pods.go:141] WaitForControlledPodsRunning: namespace(test-u3fcbg-30), labelSelector(name=small-rc-39): Pods: 0 out of 0 created, 0 running, 0 pending scheduled, 0 not scheduled, 0 inactive, 1 terminating, 0 unknown, 0 runningButNotReady                        
W0702 17:26:53.945] I0702 17:26:53.945227   14684 wait_for_pods.go:141] WaitForControlledPodsRunning: namespace(test-u3fcbg-30), labelSelector(name=small-rc-39): Pods: 0 out of 0 created, 0 running, 0 pending scheduled, 0 not scheduled, 0 inactive, 0 terminating, 0 unknown, 0 runningButNotReady                        
W0702 19:22:18.865] W0702 19:22:18.865645   14684 reflector.go:303] PodStore: namespace(test-u3fcbg-30), labelSelector(name=small-rc-39): watch of *v1.Pod ended with: too old resource version: 24914280 (25267175)                                                                                                           
W0702 19:57:06.503] E0702 19:57:06.501822   14684 clusterloader.go:181]   Errors: [measurement call WaitForControlledPodsRunning - WaitForRunningRCs error: 1 objects timed out: ReplicationControllers: test-u3fcbg-30/small-rc-39]
```

It seems RC has timed out creating pod store, but was operating just fine for the rest of the load test. 

```
$ fgrep 1512726023 build-log.txt
W0702 14:34:15.614] I0702 14:34:15.612925   14684 trace.go:81] Trace[1512726023]: "Reflector PodStore: namespace(test-u3fcbg-30), labelSelector(name=small-rc-39) ListAndWatch" (started: 2019-07-02 14:32:05.393780229 +0000 UTC m=+22640.906973464) (total time: 2m10.21911428s):                                           
W0702 14:34:15.614] Trace[1512726023]: [2m10.219073826s] [2m10.219073826s] Objects listed
```

Based on the trace it seems, that pod listing took 2m10s - 10s more than deadline. Still quite a lot. 

However based on the api_responsiveness measurements, we simply got unlucky here. Both measurements didn't indicate any problems with LIST pod calls - latencies are withing SLO threshold.
```
 151020 W0702 19:56:47.131] I0702 19:56:47.131362   14684 api_responsiveness_prometheus.go:89] APIResponsiveness: Top latency metric: {Resource:replicationcontrollers Subresource: Verb:LIST Scope:cluster Latency:{Perc50:5.5s Perc90:5.9s Perc99:5.99s} Count:3}; threshold: 10s
 151021 W0702 19:56:47.132] I0702 19:56:47.131409   14684 api_responsiveness_prometheus.go:89] APIResponsiveness: Top latency metric: {Resource:nodes Subresource: Verb:LIST Scope:cluster Latency:{Perc50:93.305476ms Perc90:950ms Perc99:4.174999999s} Count:7956}; threshold: 10s
 151022 W0702 19:56:47.132] I0702 19:56:47.131418   14684 api_responsiveness_prometheus.go:89] APIResponsiveness: Top latency metric: {Resource:pods Subresource: Verb:LIST Scope:namespace Latency:{Perc50:2s Perc90:3s Perc99:3.451749999s} Count:47770}; threshold: 5s
 151023 W0702 19:56:47.133] I0702 19:56:47.131425   14684 api_responsiveness_prometheus.go:89] APIResponsiveness: Top latency metric: {Resource:services Subresource: Verb:LIST Scope:cluster Latency:{Perc50:204.374999ms Perc90:1.667499999s Perc99:1.741749999s} Count:406}; threshold: 10s
 151024 W0702 19:56:47.133] I0702 19:56:47.131437   14684 api_responsiveness_prometheus.go:89] APIResponsiveness: WARNING Top latency metric: {Resource:replicationcontrollers Subresource: Verb:PATCH Scope:namespace Latency:{Perc50:42.371052ms Perc90:447.899188ms Perc99:1.626316666s} Count:16400}; threshold: 1s
 151025 W0702 19:56:47.133] I0702 19:56:47.131452   14684 api_responsiveness_prometheus.go:89] APIResponsiveness: WARNING Top latency metric: {Resource:leases Subresource: Verb:GET Scope:namespace Latency:{Perc50:27.888116ms Perc90:53.867424ms Perc99:1.341191893s} Count:17497474}; threshold: 1s
 151026 W0702 19:56:47.133] I0702 19:56:47.131468   14684 api_responsiveness_prometheus.go:89] APIResponsiveness: WARNING Top latency metric: {Resource:nodes Subresource:status Verb:PATCH Scope:cluster Latency:{Perc50:26.745769ms Perc90:48.142384ms Perc99:1.22328288s} Count:5836956}; threshold: 1s
 151027 W0702 19:56:47.134] I0702 19:56:47.131481   14684 api_responsiveness_prometheus.go:89] APIResponsiveness: WARNING Top latency metric: {Resource:replicationcontrollers Subresource: Verb:POST Scope:namespace Latency:{Perc50:27.753371ms Perc90:51.031875ms Perc99:1.208899999s} Count:16400}; threshold: 1s
 151028 W0702 19:56:47.134] I0702 19:56:47.131492   14684 api_responsiveness_prometheus.go:89] APIResponsiveness: WARNING Top latency metric: {Resource:configmaps Subresource: Verb:POST Scope:namespace Latency:{Perc50:29.411764ms Perc90:150ms Perc99:1.199999999s} Count:2348}; threshold: 1s
 151029 W0702 19:56:47.134] I0702 19:56:47.131506   14684 api_responsiveness_prometheus.go:89] APIResponsiveness: WARNING Top latency metric: {Resource:serviceaccounts Subresource: Verb:GET Scope:namespace Latency:{Perc50:26.136363ms Perc90:47.045454ms Perc99:1.127499999s} Count:5376}; threshold: 1s
 151030 W0702 19:56:47.135] I0702 19:56:47.131518   14684 api_responsiveness_prometheus.go:89] APIResponsiveness: WARNING Top latency metric: {Resource:pods Subresource: Verb:DELETE Scope:namespace Latency:{Perc50:29.751771ms Perc90:95.831495ms Perc99:1.030307142s} Count:333123}; threshold: 1s
 151031 W0702 19:56:47.283] I0702 19:56:47.282278   14684 api_responsiveness.go:136] APIResponsiveness: Top latency metric: {Resource:replicationcontrollers Subresource: Verb:LIST Scope:cluster Latency:{Perc50:2.259ms Perc90:8.646584s Perc99:8.646584s} Count:4}; threshold: 10s
 151032 W0702 19:56:47.283] I0702 19:56:47.282345   14684 api_responsiveness.go:136] APIResponsiveness: Top latency metric: {Resource:pods Subresource: Verb:LIST Scope:namespace Latency:{Perc50:45.103ms Perc90:100.92ms Perc99:1.899222s} Count:47770}; threshold: 5s
 151033 W0702 19:56:47.284] I0702 19:56:47.282354   14684 api_responsiveness.go:136] APIResponsiveness: Top latency metric: {Resource:nodes Subresource: Verb:LIST Scope:cluster Latency:{Perc50:75.029ms Perc90:497.518ms Perc99:1.459275s} Count:7960}; threshold: 10s
 151034 W0702 19:56:47.284] I0702 19:56:47.282363   14684 api_responsiveness.go:136] APIResponsiveness: Top latency metric: {Resource:services Subresource: Verb:LIST Scope:cluster Latency:{Perc50:112.401ms Perc90:197.054ms Perc99:1.08214s} Count:411}; threshold: 10s
 151035 W0702 19:56:47.285] I0702 19:56:47.282370   14684 api_responsiveness.go:136] APIResponsiveness: Top latency metric: {Resource:daemonsets Subresource:status Verb:PUT Scope:namespace Latency:{Perc50:3.655ms Perc90:379.804ms Perc99:379.804ms} Count:8}; threshold: 1s
```

Interestingly though, there are discrepancies between SLI measured old (api_responsiveness.go) and new way (api_responsiveness_prometheus.go). The later is closer to SLO definition and more strict than former. I think we should start to use the newer one to fail test if needed (currently it only does calculations).

**Comment by oxddr** [comment]
/close 

This seems to be flake :(

**Comment by wojtek-t** [comment]
I think we should figure out what to do to avoid this flake in the future instead of closing this.

**Comment by oxddr** [comment]
I was about to open a separate issue to track this, but let's reuse this issue then. 

@mborsz - you mentioned you had a similar problem in the past. Any chance for any pointers?

**Comment by oxddr** [comment]
I've checked apiserver logs around that time and it seems that latency of this call on its side were order of tens of miliseconds
```
$ zgrep small-rc-39 kube-apiserver.log-20190702-1562082302.gz | fgrep test-u3fcbg-30  | fgrep clusterloader
I0702 14:34:15.214533       1 wrap.go:47] GET /api/v1/namespaces/test-u3fcbg-30/pods?labelSelector=name%3Dsmall-rc-39&limit=500&resourceVersion=0: (53.7946ms) 200 [clusterloader/v0.0.0 (linux/amd64) kubernetes/$Format 35.232.62.47:43304]
```

I've checked also another trace from around that time, with long listing time:
```
$ fgrep Trace build-log.txt | egrep 14:3 
W0702 14:34:15.613] I0702 14:34:15.612750   14684 trace.go:81] Trace[986770188]: "Reflector PodStore: namespace(test-u3fcbg-44), labelSelector(name=small-rc-215) ListAndWatch" (started: 2019-07-02 14:32:32.305024758 +0000 UTC m=+22667.818217980) (total time: 1m43.307686927s):                                           
W0702 14:34:15.613] Trace[986770188]: [1m43.30763575s] [1m43.30763575s] Objects listed                                      
```
and it shows similar symptoms (long listing time on the client side, short on the apiserver side)
```
$ zgrep small-rc-215 kube-apiserver.log-20190702-1562082302.gz | fgrep test-u3fcbg-44  | fgrep clusterloader                                                                                                                                                           
I0702 14:34:15.207872       1 wrap.go:47] GET /api/v1/namespaces/test-u3fcbg-44/pods?labelSelector=name%3Dsmall-rc-215&limit=500&resourceVersion=0: (47.184101ms) 200 [clusterloader/v0.0.0 (linux/amd64) kubernetes/$Format 35.232.62.47:43304]                                                                              
```

Current hypothesis is that it's actually not a flake, but rather symptom of a packet drop that we've started to observe earlier. @mborsz is investigating it.

**Comment by oxddr** [comment]
My understanding is that, due to packet drop TCP retries causing visible latency on the client side, while apiserver's latency is fine, because it's logged before actual packets are sent. In the end client is able to get all packets, but it takes time. @mborsz may be able to provide more details - that's his hypothesis. 

There were no retries (no trace of them in build-log and apiserver logs).

**Comment by wojtek-t** [comment]
I'm not sure it's exactly true. We're logging that the request finished after Write() operation to responseWriter is finished. And my understand was that it is actually a synchronous operations (we've seen a bunch of cases in the past that slow client was slowing down that operation).
But maybe I'm missing something

**Comment by alenkacz** [comment]
@oxddr @wojtek-t any update on this one?

**Comment by oxddr** [comment]
This is currently blocked by investigation by @mborsz on a packet drop we observe in large tests. As mentioned earlier in this issue, we suspect this may be causing long latency observed on the client side.
